### PR TITLE
[P2] Onboarding Clarity: handler-layer UX + selfcheck (supersedes #1261, #866)

### LINKED ISSUE #326: [P2] Improve onboarding experience
Improve onboarding experience: 

- [ ] Clearer logging when node is launched and waiting for PoC
- [ ] remove errors which doesn’t mean errors
- [ ] automatic testing that everything will work when PoC starts (models can be deployed, all endpoints are accessible); 
- [ ] Clean up logs to avoid confusing ERROR messages;

Description of the proposal: 
[https://github.com/gonka-ai/gonka/blob/a2a15267ea4aa55288fc873f4e5e68bc69366447/proposals/onboarding-clarity-v1/README.md](https://www.google.com/url?q=https://github.com/gonka-ai/gonka/blob/a2a15267ea4aa55288fc873f4e5e68bc69366447/proposals/onboarding-clarity-v1/README.md&sa=D&source=docs&ust=1756918029145609&usg=AOvVaw3j8xsw5yvYWpDeCcVHBhZY)

### PR DESCRIPTION
**Pull request by bonujel**
## Summary

Supersedes #866, completes #1261. Implements #326 (clearer onboarding logs, no false "errors", pre-PoC readiness testing). Onboarding UX is derived read-only at the **admin handler layer**, and the broker's reaction to PoC phases is verified in a separate `api selfcheck` mode against a mocked chain. 

broker.NodeState is unchanged — based on [Dima’s](https://github.com/gonka-ai/gonka/pull/866#issuecomment-4172544143) and [patimen’s](https://github.com/gonka-ai/gonka/pull/866#issuecomment-4167136354) comments in #866, the broker should not introduce any state changes, so the only broker change here is a behavior-preserving refactor.

- **Onboarding status** — `GET /admin/v1/nodes` returns an optional `onboarding` block: MLnode state `WAITING_FOR_POC`/`TESTING`/`TEST_FAILED`, participant `INACTIVE_WAITING`/`ACTIVE_PARTICIPATING`, timing, and human messages. Derived read-only (`TimingCalculator`/`StatusReporter`/`OnboardingStateManager` + a 30s-cached `ActivityTracker`, no per-request chain RPCs).
- **Pre-PoC testing** — `POST /admin/v1/nodes/:id/test` (manual) plus **event-driven auto-test**: the block dispatcher invokes an admin hook after each synced epoch update, which retries pre-PoC validation for broker-registered nodes (covering nodes loaded from config at startup, or whose first test was skipped near PoC). Runs model load → health → inference request and records timings. Pass → `WAITING_FOR_POC` (the "safe to be offline" wording only after a pass); fail → `TEST_FAILED`, logged at the API node. `TEST_FAILED` comes only from the test, never from the broker's operational FAILED.
- **`api selfcheck`** — drives the real broker through synthetic PoC phases (`PoCGenerate → PoCValidate → Inference`) and asserts its **intended-state transitions** per stage, exit 0/1. It verifies broker PoC-phase wiring at the intended-state level; it does not exercise a real MLnode worker.
- **Log cleanup** — a "waiting for PoC (starts in 2h 15m)…" line at startup and every 5 min while onboarding; benign startup/inactive logs demoted to Info; genuine failures stay visible.

## Broker: behavior-preserving refactor for test fidelity

The pre-PoC test must bring an MLnode up **exactly the way the broker does at PoC**, or it reports false pass/fail. Rather than re-implement that path in the test (and let it drift from the broker), we extracted the broker's model-launch logic into pure functions the test reuses:

- `broker.BuildModelLaunchPlan` / `BuildConfiguredModelLaunchPlans` / `StopBeforeModelLaunch`
- `broker.MergeModelArgs` (governance + local args), and one `apiconfig.MLNodeURL` for versioned URLs

This is the only change to the broker, and it is **behavior-preserving**: no `NodeState` fields, no new broker commands — just a new pure-function file plus delegations from `InferenceUpNodeCommand`. Covered by the broker test suite and `api selfcheck`.

## Known limitation: best-effort test safety (no broker test-lease)

Per @patimen and @DimaOrekhovPS on #866, we add **no state or locks to the broker**. As a consequence, the check that protects assigned/preserved/PoC nodes before a test is a **point-in-time snapshot**: the test then loads and stops the MLnode outside the broker's state machine for up to a few minutes, so a node assigned or preserved *during* the test could be stopped mid-work — a known corner case.

We accept this rather than introduce a broker-level "reserve-if-idle" test lease (which would mean exactly the broker state/complexity #866 was asked to avoid). The window is kept small and the impact bounded:

- Auto-tests run only on nodes that look idle and only when there is >1h until PoC.
- The event-driven hook fires **after** the broker refreshes per-epoch node assignment (`EpochMLNodes`/`PreservedModels`), so a just-assigned node is not mistaken for idle.
- The broker's next-block reconciliation re-brings-up any node a test left stopped.

Fully closing the window would require an atomic broker-side reservation; that is intentionally out of scope here.

## Why not #866

#866 added UX state to `broker.NodeState` + new broker commands; reviewers (@patimen, @DimaOrekhovPS) flagged this as complexity on a critical path. Same features, broker state untouched:

| | #866 | This PR |
|---|---|---|
| `broker.NodeState` new fields | 6 | 0 |
| broker new commands | 2 | 0 |
| per-request chain RPCs in `GET /nodes` | N+1 | 0 (30s cache) |
| UX state computation | broker writes back | handler derives |
| broker change | feature logic | behavior-preserving refactor only |

Product context in `proposals/onboarding-clarity-v1/README.md`.

## Test plan

- [x] `go build` / `go vet` / `gofmt` / `go test` (broker, admin, selfcheck, event_listener, participant, mlnodeclient) — green
- [x] `api selfcheck` — 6 stages PASS, exit 0
- [x] Verified natively (built from source): `api selfcheck` 6/6 and the changed-package test suites green
- [ ] Against a **real MLnode (vLLM)** / the **public testnet** — not yet run

### CODE REVIEW

**Comment by bonujel** [comment]
@codex review

**Comment by bonujel** [comment]
@tcharchian could you take a look at this when you have a chance? 🙏

This PR improves the node onboarding experience (#326), reworked around the direction Patiment and Dima gave on #866 — keeping onboarding UX out of the broker's critical path and deriving it in the admin layer instead, rather than adding UX state/commands to the broker.

What it does:

- Onboarding envelope on `GET /admin/v1/nodes`— a read-only `onboarding` block (participant state, MLnode state, PoC timing, a human-facing message/guidance) derived from broker output, chain phase, cached participant activity, and the latest test result. Nothing is written back to the broker.

- Manual MLnode test — `POST /admin/v1/nodes/:id/test` validates a node's configured models the way the broker would launch them (stop → inference-up → health → real inference probe), with safety checks so it won't disrupt a node that's serving, assigned, or near PoC.

- Pre-PoC auto-test — new/changed nodes, and nodes re-checked on each synced block, get auto-tested when there's enough time before the next PoC, so config problems surface early instead of at PoC time.

- Clearer logging — a periodic onboarding status line ("waiting for PoC", etc.) and demoting benign "no epoch models" noise for participants that aren't active yet.

- `api selfcheck` — a standalone mode that drives the real broker against a mocked chain to verify PoC-phase wiring offline, no real MLnode or testnet required.

Build, `go vet`, and the targeted test suites pass, and `api selfcheck` reports a clean `PASS`. Happy to walk through any part.

**Comment by patimen** [comment]
@bonujel If I may offer an idea:
What if we leaned a bit more toward composability here?

I wonder if, instead of putting more onboarding-specific state/behavior into the runtime, we could expose a few small factual queries/actions and let the onboarding flow be composed on top of them, either by a human operator or by an AI `skill.md`

A small example from this PR would be the difference between:
- teaching the system a richer onboarding state machine like TESTING / TEST_FAILED / auto-retry / “safe to be offline”
vs.
- exposing simpler facts like “what nodes are configured?”, “what does the broker currently think about them?”, “how long until PoC?”, or “what launch plan would this node use right now?”

Then the higher-level onboarding guidance becomes something layered on top of those primitives, instead of baked into the server itself.
My instinct is that this would give us something more flexible and reusable, without committing the runtime to as much onboarding-specific logic.

**Comment by bonujel** [comment]
@patimen Thanks, this is a really helpful framing — I went ahead and leaned into it.

One thing that might not be obvious from the diff: the onboarding block is already a read-only derivation (nothing is written back to the broker or kept as runtime state), so the state names and wording are just a thin convenience layer over the facts — removable later without any migration.

To make the underlying primitives first-class, I've added three small read-only fact endpoints (built from pieces the PR already introduces, no new runtime state). They line up with the questions you raised:

| Primitive (fact / action) | Endpoint |
|---|---|
| What nodes are configured / what does the broker currently think about them? | `GET /nodes` (raw broker view; the `onboarding` block is derived on top of it) |
| How long until PoC? | `GET /poc/timing` **(new)** — phase, blocks/seconds until next PoC; `available:false` until synced |
| What launch plan would this node use right now? | `GET /nodes/:id/launch-plan` **(new)** — final merged args per model, plus configured models that would *not* launch and why; same selection pipeline as the broker, so the preview cannot drift |
| Run a readiness test / read the raw result back | `POST /nodes/:id/test` / `GET /nodes/:id/test` **(GET new)** — raw `TestResult` with per-model load times, failing model, retryability |
| Is the environment set up correctly? | `GET /setup/report` |

On the derived layer — I'd lean toward keeping it in this PR for now, mostly because of how #326 is scoped: it asks the node itself to surface "waiting for PoC" as the most visible log and to auto-test new MLnodes, and those two have to run in-process. The default wording is really just to give operators something clear out of the box. Happy to revisit how thick that layer should be.

The natural next step is exactly the composition layer you describe — a skill.md that derives the guidance from these primitives. Once that proves out, the wording layer here can be thinned or dropped without migration, since nothing behind it is stored. Does that split sound reasonable to you?

**Comment by bonujel** [comment]
@patimen Just a gentle nudge on this one. I've pushed the three read-only fact endpoints we discussed. Whenever you have a moment, I'd really appreciate your thoughts on the approach and a review. Happy to adjust anything. Thanks!
cc @tcharchian

**Comment by bonujel** [comment]
Hi @tcharchian, gentle ping on #1296, I've updated the PR based on earlier review feedback. 
Do you think the changes look acceptable? Would appreciate another look when you get a chance, so that we can move forward with #1359 that depends on this PR. Thanks!

**Comment by tcharchian** [comment]
Hi @bonujel, I believe @DimaOrekhovPS will get back to you on this shortly.

**Comment by bonujel** [comment]
Rebased onto upgrade-v0.2.16 to match the milestone (the old upgrade-v0.2.14 base is no longer shipping), could you please take a look when you have a moment? @tcharchian, @DimaOrekhovPS

File: decentralized-api/broker/broker.go [hunk 1]
```diff
@@ -340,6 +341,17 @@ func NewBroker(chainBridge BrokerChainBridge, phaseTracker *chainphase.ChainPhas
...
+// Stop shuts down the broker's background workers. Idempotent and safe to call
+// from any goroutine.
+//
+// Contract after Stop: QueueMessage returns ErrBrokerStopped rather than
+// accepting a command nobody will execute, and GetNodes returns that error
+// instead of blocking. Commands already queued are not drained, so a caller that
+// needs a command's result must await it before stopping.
+func (b *Broker) Stop() {
+	b.stopOnce.Do(func() { close(b.stop) })
+}
```
[the remark this answers is not in the document]
Comment (bonujel) [reply decentralized-api/broker/broker.go:364]: Fixed. 
Stop() now shuts down the per-node workers too, not just the broker-level loops.

File: decentralized-api/internal/event_listener/new_block_dispatcher.go [hunk 2]
```diff
@@ -177,6 +187,12 @@ func NewOnNewBlockDispatcherFromCosmosClient(
 	return dispatcher
 }
 
+// SetOnEpochState registers a hook invoked on every new block once the chain
+// is synced, with the freshly-updated epoch state.
+func (d *OnNewBlockDispatcher) SetOnEpochState(hook EpochStateHook) {
+	d.onEpochState.Store(&hook)
+}
```
[the remark this answers is not in the document]
Comment (bonujel) [reply decentralized-api/internal/event_listener/new_block_dispatcher.go:201]: Fixed. Clear the stored pointer instead of storing &hook for a nil func, so the load site's non-nil check can no longer pass and then invoke a nil function. 
Added a test covering the nil path.

File: decentralized-api/internal/server/admin/mlnode_tester.go [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+	result := t.runOnce(runCtx, cfg, handle)
+	result.Fingerprint = fingerprint
+
+	// A cancelled run says nothing about the node: either the caller went away
+	// or the node was reconfigured mid-test (Invalidate). Recording its failure
+	// would latch a bogus TEST_FAILED and, worse, describe inputs that no
+	// longer exist. Report it to the caller but keep it out of the record.
+	if ctxErr := runCtx.Err(); ctxErr != nil && result.Status == TestFailed {
+		return result, nil
+	}
```
[the remark this answers is not in the document]
Comment (bonujel) [reply decentralized-api/internal/server/admin/mlnode_tester.go:313]: Fixed. now skip recording only on context.Canceled (caller gone / Invalidate). A DeadlineExceeded run is recorded as a retryable failure, so auto-test's backoff engages instead of re-testing the node on every synced block. Added a test asserting a timed-out run is recorded as retryable.

**Comment by bonujel** [comment]
> ## Pull request overview
> This PR improves node onboarding clarity by deriving UX-facing onboarding state at the **admin handler layer** (read-only), adding pre-PoC MLnode validation (manual + event-driven auto-test), and introducing an `api selfcheck` mode that drives the real broker through synthetic PoC phases against a mocked chain.
> 
> **Changes:**
> 
> * Add derived onboarding envelope to `GET /admin/v1/nodes`, plus periodic “waiting for PoC” logging and read-only “fact” endpoints.
> * Add MLnode validation testing (`POST /admin/v1/nodes/:id/test`) with event-driven auto-test retried after synced epoch updates.
> * Add `api selfcheck` subcommand and broker refactors/pure helpers to ensure test fidelity and avoid behavior drift.
> 
> ### Reviewed changes
> Copilot reviewed 48 out of 48 changed files in this pull request and generated 3 comments.
> 
> Show a summary per file

All 3 issues above are fixed.

**Comment by bonujel** [comment]
Bugs comment by copilot has been addressed. I also ran a local review with Codex and didn’t identify any additional actionable issues.
When you have a chance, could you please continue with the review? Thanks! @DimaOrekhovPS, cc @tcharchian

**Comment by bonujel** [comment]
Starting to break this up. At 51 files it is too large to review as one change, and it has been open long enough that the code underneath it has moved twice — the vLLM 0.25.1 work in #1536 landed `ResolveModelDeployment`, which overlaps the broker refactor at the base of this branch.

Rather than rebase the whole thing, I am slicing it into PRs that can be reviewed on their own, the same way #1518 came out of it earlier. This PR stays open as the umbrella so the overall shape is still visible.

First slice: #1644 — the `decentralized-api selfcheck` subcommand. Self-contained, `+670/-0`, built directly on current `upgrade-v0.2.16` so it uses `ResolveModelDeployment` from the start rather than the older helper on this branch.

Planned slices after that:

- **pre-PoC MLnode test** (manual and event-driven). Needs the broker helper on this branch replaced by `ResolveModelDeployment` first — four call sites.
- **onboarding state and read-only fact endpoints** — the largest remaining piece.

The `fix(mlnodeclient)` commit here is already covered by #1518 and will not be sliced again.
